kotlin/src/main/java/com/squareup/moshi/KotlinJsonAdapter.kt

```diff
@@ -0,0 +1,141 @@
+  override fun create(type: Type, annotations: Set<Annotation>, moshi: Moshi): JsonAdapter<*>? {
+    val kclass = (type as? Class<*>)?.kotlin ?: return null
+    if (!kclass.isData) return null
+    // This gives us a few guarantees:
+    // 1. Primary constructor exists
+    // 2. Constructor params are var/val and have backing fields
+    // 3. Class is not abstract or an interface
+    // 4. Class is final
+
+    val constructor = kclass.primaryConstructor!!
+    constructor.isAccessible = true
+
+    val concretePropertyByName = kclass.declaredMemberProperties
+        .filter { it.javaField != null }
+        .associateBy { it.name }
+    val constructorParameterByName = constructor.parameters.associateBy { it.name }
+    for ((name, property) in concretePropertyByName) {
+      if (name !in constructorParameterByName.keys &&
+          !Modifier.isTransient(property.javaField!!.modifiers)) {
+        throw IllegalArgumentException("Property $name is not transient. " +
+            "Any properties not found in the primary constructor need to be transient. " +
+            "If this is a delegated property, mark the delegate as transient " +
+            "using @delegate:Transient.")
+      }
+    }
+
+    constructor.parameters
+        .asSequence()
+        .filterNot { it.isOptional }
+        .map { concretePropertyByName[it.name]!! }
+        .firstOrNull { it.findAnnotation<Transient>() != null }
+        ?.let { throw IllegalArgumentException("Transient properties in primary constructor are unsupported. ($it)") }
+
+    val isPlatformType = isPlatformType(Types.getRawType(type))
+    val notIncludedParams = constructorParameterByName
+        .filterNot { it.value.isOptional }
+        .keys.map { concretePropertyByName[it]!! }
+        .filter { !includeField(isPlatformType, it.javaField!!.modifiers) }
+    if (notIncludedParams.isNotEmpty()) {
+      throw IllegalArgumentException("Primary constructor contains parameters " +
+          "not found in serialized JSON: $notIncludedParams")
+    }
+
+    val toJsonAdapter = ClassJsonAdapter.FACTORY.create(type, annotations, moshi)
+    val propertyByParam = constructor.parameters.associate { it to concretePropertyByName[it.name]!! }
+    val adapterByParam = constructor.parameters.associate { param ->
+      val javaField = propertyByParam[param]!!.javaField!!
+      val paramType = Types.resolve(javaField.type, Types.getRawType(javaField.type), javaField.genericType)
+      val annotations = param.annotations.toSet()
+
+      param to moshi.adapter<Any>(paramType, annotations)
+    }
+
+    return KotlinJsonAdapter(constructor, toJsonAdapter as JsonAdapter<Any?>, propertyByParam, adapterByParam)
+  }
+
+}
+
+@Suppress("LoopToCallChain")
+internal class KotlinJsonAdapter<T>(
+    private val constructor: KFunction<T>,
+    private val toJsonAdapter: JsonAdapter<T>,
+    private val propertyByParam: Map<KParameter, KProperty1<out Any, Any?>>,
```

Comment:
Ideally for performance we avoid map lookups in the decode/encode paths. Otherwise this will be slower than Java and nobody wants that